### DISCUSSION TITLE
compile_error!("Only features sync,macros,io-util,rt,time are supported on wasm.");

### DISCUSSION THREAD

**Discussion by otaviomacedo** [winglang/wing#749] [Q&A] [answered]
When I run `npm run build`, it fails with:

    ERROR: Something went wrong in run-commands - Command failed: cargo wasi build

Running `cargo wasi build` individually, I get:

```
   Compiling wingii v0.1.0 (/Users/otaviom/projects/wing/libs/wingii)
   Compiling tokio v1.21.1
error: Only features sync,macros,io-util,rt,time are supported on wasm.
   --> /Users/otaviom/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/lib.rs:462:1
    |
462 | compile_error!("Only features sync,macros,io-util,rt,time are supported on wasm.");
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: could not compile `tokio` due to previous error
```

I've tried changing the `features` attribute of the `tokio` package to `features = ["sync","macros","io-util","rt","time"]`. But then I get another compilation error (which sorta makes sense):

```
error[E0425]: cannot find function `stdout` in module `tokio::io`
          --> apps/wing-language-server/src/main.rs:306:26
           |
       306 |     let stdout = tokio::io::stdout();
           |                             ^^^^^^ not found in `tokio::io`
           |
       help: consider importing this function
           |
       1   | use std::io::stdout;
           |
       help: if you import `stdout`, refer to it directly
           |
       306 -     let stdout = tokio::io::stdout();
       306 +     let stdout = stdout();
           |
```

I'm not familiar with Rust, so I'm probably doing something wrong. Any help would be appreciated!

---

Dev environment:

```
npm version: 8.19.1
rustc 1.65.0 (897e37553 2022-11-02)
cargo 1.65.0 (4bc8f24d3 2022-10-20)
System Version: macOS 12.6.1 (21G217)
Kernel Version: Darwin 21.6.0
```

**Comment by staycoolcall911** [comment]
Hey, did you run `sudo bash scripts/setup_wasi.sh`? It's one of the installation phases of https://docs.winglang.io/contributors/handbook#-how-do-i-build-wing-locally

**Reply by otaviomacedo** [reply]
Yes,I did. I forgot to add that to the question.

**Reply by staycoolcall911** [reply]
Thanks, I'm able to reproduce. Working on it.

**Comment by yoav-steinberg** [comment]
`cargo wasi build` on the project root fails because not all workspace members (specifically `apps/wing-language-server`) can build to the wasi target. This is fine since we don't need the lang-server on wasi.

Regarding why `npm run build` fails, can you send more info, what appears above:
`ERROR: Something went wrong in run-commands - Command failed: cargo wasi build`

**Reply by otaviomacedo** [reply]
```
    ✖  nx run wingc:build-wasm
       Compiling thiserror v1.0.36
          Compiling wingii v0.1.0 (/Users/otaviom/projects/wing/libs/wingii)
       error[E0554]: `#![feature]` may not be used on the stable release channel
          --> /Users/otaviom/.cargo/registry/src/github.com-1ecc6299db9ec823/thiserror-1.0.36/src/lib.rs:236:34
           |
       236 | #![cfg_attr(provide_any, feature(provide_any))]
           |                                  ^^^^^^^^^^^
       
       For more information about this error, try `rustc --explain E0554`.
       error: could not compile `thiserror` due to previous error
       warning: build failed, waiting for other jobs to finish...
       
        >  NX   ERROR: Something went wrong in run-commands - Command failed: cargo wasi build
       
          Pass --verbose to see the stacktrace.
       
```

**Reply by yoav-steinberg** [reply]
seems to be related to the rust toolchain version you're using, can you post the output of:
`rustc --version`
and of
`rustup show`

**Reply by otaviomacedo** [reply]
```
$ rustup show
Default host: x86_64-apple-darwin
rustup home:  /Users/otaviom/.rustup

installed toolchains
--------------------

stable-x86_64-apple-darwin (default)
nightly-x86_64-apple-darwin

installed targets for active toolchain
--------------------------------------

wasm32-wasi
x86_64-apple-darwin

active toolchain
----------------

stable-x86_64-apple-darwin (default)
rustc 1.65.0 (897e37553 2022-11-02)
```

**Reply by yoav-steinberg** [reply]
and `rustc --version`?

**Reply by otaviomacedo** [reply]
It's the last line of the output above: `rustc 1.65.0 (897e37553 2022-11-02)`

**Comment by Chriscbr** [comment, marked as answer]
@otaviomacedo I found this GItHub issue reporting a similar problem: https://github.com/dtolnay/thiserror/issues/192. Two ideas:

- Can you try running `cargo clean` and see if that makes a difference?
- Can you check if you are using rust-analyzer (e.g. through an IDE extension) and whether it's on a recent version?

**Reply by otaviomacedo** [reply]
`cargo clean` solved the problem. Thanks, Chris!

**Reply by staycoolcall911** [reply]
Nice! Thank you @Chriscbr

**Reply by Veercodeprog** [reply]
not working for me

**Reply by Veercodeprog** [reply]
i get the error on trunk build 
473 | compile_error!("Only features sync,macros,io-util,rt,time are supported on wasm.");

**Reply by eladb** [reply]
@Veercodeprog is this working or not?

**Reply by Veercodeprog** [reply]
@eladb not working , i am trying to build using trunk specifically for something that use csr only , is it related to ssr or csr thing?

**Reply by skyrpex** [reply]
@Veercodeprog can you share the actual steps you do to get such an error? Like, what you run and in which directory exactly, also maybe share your rust toolchain version 🙏🏻
